### DISCUSSION TITLE
inherit sklearn BaseEstimator

### DISCUSSION THREAD

**Discussion by remiadon** [scikit-mine/scikit-mine#137] [General]
#### Describe the workflow you want to enable
When I was working on the project I chose not to inherit sklearn BaseEstimator's class, and instead define base classes and interfaces within this project (while still implementing tests for compatibility with sklearn).
The rationale at this time was to keep minimal dependency for this project (numpy/roaringbitmap/pandas/joblib).
With more recent contributions the list of dependencies grew quickly, now including `networkx`

I do think now sklearn should be dependency for skmine, which would encourage implementing more methods in the `feature_extraction` module (use case = skmine as preprocessing for sklearn)

#### Describe your proposed solution
- make skmine.base.BaseMiner a pure interface, and remove sklearn-duplicated code inside like _get_tags, etc ..
- include a stable version of sklearn in the requirements
- Have the main classes like itemsets.SLIM to inherit BaseEstimator, while classes like itemsets.LCM should inherit BaseTransformer 
- remove https://github.com/scikit-mine/scikit-mine/blob/master/.github/check_sklearn_compat.py, and instead implements compatibility check using sklearn itself, as in https://github.com/scikit-learn-contrib/project-template/blob/master/skltemplate/tests/test_common.py

#### Additional context
great resource here https://github.com/scikit-learn-contrib/project-template

NB.
I still think dependencies should be added with great care. In the context of this project `networkx` should be optional

**Comment by thomasbtnfr** [comment]
Hello @remiadon,
Thanks for your feedback, it's an interesting point. This will be discussed with the team soon to determine the priorities to implement.

**Comment by hermann74** [comment]
Currently, we have transferred LCM, SLIM and associated estimators to the sklearn databases. The choice of the TransformerMixin had the effect of renaming the discover method to transform. 

See branch sklearn_fully_compatible for change in code .

Concerning the compatibility, or rather the validation by check_estimator, it was not valid before (if we remove the excludes tags on the estimators, and the filtering of the estimators) and is not valid now.
The blocking point being the inputs ! Indeed sklearn presupposes 2d array of continuous data (Xtypes='2darray') . The other types 'categorical', '2dlabels' are not supported in the tests, but will be perhaps in the next versions.

**Comment by hermann74** [comment]
in version 1.0, LCM , SLIM and PeriodicPatternMiner inherit from sklearn bases : BaseEstimator and TransformerMixin
- this brings automatic tests but those concerning the type of our inputs (categorical) will soon be taken into account in sklearn 
- for these 3 estimators, the estimation on the dataset D is done in the `estim.fit(D)` , but then in the transform call `estim.transform(D)` , the argument D is not used.

In fact, they are not transformers in the sense that once fitted, they transform an X into a Z with the same number of lines 
Using sklearn.BaseEstimator and skmine.DiscoverMixin would probably be a better compromise
